.ci/go-test.sh

```diff
@@ -0,0 +1,109 @@
+#!/bin/bash -e
+#
+# Copyright (c) 2017-2018 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+# Set default test run timeout value.
+#
+# KATA_GO_TEST_TIMEOUT can be set to any value accepted by
+# "go test -timeout X"
+timeout_value=${KATA_GO_TEST_TIMEOUT:-10s}
+
+# KATA_GO_TEST_FLAGS can be set to change the flags passed to "go test".
+go_test_flags=${KATA_GO_TEST_FLAGS:-"-v -race -timeout $timeout_value"}
+
+test_packages=$(go list ./... 2>/dev/null || true)
+
+test_coverage_file="profile.cov"
+cov_mode=0644
+tmp_coverage_file="profile-tmp.cov"
+html_report_file="coverage.html"
+
+# Run a command as either root or the current user (which might still be root).
+#
+# If the first argument is "root", run using sudo, else run as the current
+# user. All arguments after the first will be treated as the command to run.
+run_as_user()
+{
+	user="$1"
+	shift
+	cmd=$*
+
+	if [ "$user" = root ]; then
+		# use a shell to ensure PATH is correct.
+		sudo -E PATH="$PATH" sh -c "$cmd"
+	else
+		$cmd
+	fi
+}
+
+# Run the tests and generate an HTML report of the results
+test_html_coverage()
+{
+	test_coverage
+
+	go tool cover -html="${test_coverage_file}" -o "${html_report_file}"
+	rm -f "${test_coverage_file}"
+
+	run_as_user "current" chmod "${cov_mode}" "${html_report_file}"
+}
+
+# Run all tests and generate a test coverage file.
+test_coverage()
+{
+	echo "mode: atomic" > "$test_coverage_file"
+
+	if [ $(id -u) -eq 0 ]; then
+		echo >&2 "WARNING: Already running as root so will not re-run tests as non-root user."
+		echo >&2 "WARNING: As a result, only a subset of tests will be run"
+		echo >&2 "WARNING: (run this script as a non-privileged to ensure all tests are run)."
+		users="current"
+	else
+		# Run the unit-tests *twice* (since some must run as root and
+		# others must run as non-root), combining the resulting test
+		# coverage files.
+		users="current root"
+	fi
+
+	echo "INFO: Currently running as user '$(id -un)'"
+
+	for pkg in $test_packages; do
+		for user in $users; do
+			printf "INFO: Running 'go test' as %s user on package '%s' with flags '%s'\n" \
+				"$user" "$test_packages" "$go_test_flags"
+
+			run_as_user "$user" go test $go_test_flags -covermode=atomic -coverprofile="$tmp_coverage_file" "$pkg"
+
+			if [ -f "${tmp_coverage_file}" ]; then
```

Comment:
This line feels oddly redundant, but benign, to me - maybe the mix of `tmp` and `test` prefixes is confusing me - won't this file always exist from the line above it? I suspect I'm missing something subtle?, in which case, do we think we can elaborate a touch in the script?

Reply:
That's a good question, but the test is in fact required. I've added a comment explaining why (and also caught and fixed a bug in the `printf` call above by doing that).